File: alds1/8a_binary_search_tree1.c
#define dbg(fmt,...) fprintf(stderr,fmt,__VA_ARGS__)
#define dpri(x) dbg(#x ": %d\n", x)
#define dprs(x) dbg(#x ": %s\n", x)
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
typedef long long ll;
const int MYINF = 1e9+7;
typedef int bool;
const bool true = 1;
const bool false = 0;

typedef struct node_st {
	int key;
	struct node_st *left, *right;
} node_t;

node_t *root;

node_t *node_new(int x) {
	node_t *new = malloc(sizeof(node_t));
	if (new == NULL) exit(1);
	new->key = x;
	new->left = new->right = NULL;
	return new;
}

node_t *insert(node_t *nd, int x) {
	if (!nd) return node_new(x);
	if (x < nd->key) nd->left = insert(nd->left, x);
	else if (nd->key < x) nd->right = insert(nd->right, x);
	return nd;
}

enum {
	PREORDER,
	INORDER,
};

void print_tree(node_t *nd, int order) {
	if (!nd) return;
	if (order == PREORDER) printf(" %d", nd->key);
	print_tree(nd->left, order);
	if (order == INORDER) printf(" %d", nd->key);
	print_tree(nd->right, order);
}

int main(int argc, char **argv)
{
	int i, n, x;
	char cmd[128];

	scanf("%d", &n);
	for (i = 0; i < n; i++) {
		scanf("%s", cmd);
		if (!strcmp("insert", cmd)) {
			scanf("%d", &x);
			root = insert(root, x);
		} else if (!strcmp("print", cmd)) {
			print_tree(root, INORDER);
			puts("");
			print_tree(root, PREORDER);
			puts("");
		}
	}

	return 0;
}

File: alds1/8a_binary_search_tree1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"unicode"
)

type Node struct {
	key         int
	left, right *Node
}

type Tree struct {
	root *Node
}

func NewTree() *Tree {
	return &Tree{}
}

func (t *Tree) Insert(k int) {
	y := &t.root
	x := *y
	for x != nil {
		if k < x.key {
			y = &x.left
		} else if x.key < k {
			y = &x.right
		} else {
			// already inserted
			return
		}
		x = *y
	}

	*y = &Node{
		key:   k,
		left:  nil,
		right: nil,
	}
}

const (
	Inorder = iota
	Preorder
)

func (t *Tree) Print(order int) {
	var rec func(*Node)
	rec = func(nd *Node) {
		if nd == nil {
			return
		}
		if order == Preorder {
			printf(" %d", nd.key)
		}
		rec(nd.left)
		if order == Inorder {
			printf(" %d", nd.key)
		}
		rec(nd.right)
	}
	rec(t.root)
	println()
}

func (t *Tree) PrintInorder() {
	t.Print(Inorder)
}

func (t *Tree) PrintPreorder() {
	t.Print(Preorder)
}

func main() {
	defer stdout.Flush()
	m := readInt()
	t := NewTree()
	for i := 0; i < m; i++ {
		cmd := readString()
		switch cmd {
		case "insert":
			x := readInt()
			t.Insert(x)
		case "print":
			t.PrintInorder()
			t.PrintPreorder()
		default:
			panic("unknown command")
		}
	}
}

var (
	stdin  *bufio.Reader
	stdout *bufio.Writer
)

func init() {
	stdin = bufio.NewReader(os.Stdin)
	stdout = bufio.NewWriter(os.Stdout)
}

func scanf(f string, args ...interface{}) (int, error) {
	n, err := fmt.Fscanf(stdin, f, args...)
	for {
		r, _, err := stdin.ReadRune()
		if err != nil || !unicode.IsSpace(r) {
			stdin.UnreadRune()
			break
		}
	}
	return n, err
}

func readInt() int {
	var n int
	scanf("%d", &n)
	return n
}

func readIntSlice(n int) []int {
	a := make([]int, n)
	for i := 0; i < n; i++ {
		scanf("%d", &a[i])
	}
	return a
}

func readLengthAndSlice() []int {
	n := readInt()
	return readIntSlice(n)
}

func readString() string {
	var s string
	scanf("%s", &s)
	return s
}

func printf(f string, args ...interface{}) (int, error) {
	return fmt.Fprintf(stdout, f, args...)
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdout, args...)
}
